src/mdl/holder.rs

```diff
@@ -0,0 +1,491 @@
+//! A module to handle presentation of an ISO 18013-5 mobile driving license (mDL).
+//!
+//! This module manages the session for presentation of an mDL, including generating
+//! the QR code and BLE parameters, as well as managing the types of information requested
+//! by the reader.
+//!
+//! The presentation process is begun by calling [`initialize_mdl_presentation`] on the
+//! [`Wallet`] and passing the id of the mdoc to be used as well as a UUID that the client
+//! will use for the BLE central client:
+//!
+//! ```rust
+//! let ble_uuid = Uuid::new_v4();
+//! let presentation_session = wallet.initialize_mdl_presentation("test_mdoc", ble_uuid);
+//! // Presenting software then creates a QR code with presentation_session.qr_code_uri
+//! // and sets up a BLE session using presentation_session.ble_ident. It will then
+//! // receive a list of requested fields from the reader.
+//! let requested_items = presentation_session.handle_request(request_bytes)?;
+//! // Presenting software should then confirm with the user that they wish to present
+//! // the requested information. If so...
+//! let signed_response_bytes = presentation_session.submit_response(user_permitted_items, key_id)?;
+//! // Presenting software then sends response_bytes over BLE to the reader, completing the exchange.
+//!
+
+/// Begin the mDL presentation process for the holder.
+///
+/// Initializes the presentation session for an ISO 18013-5 mDL and stores
+/// the session state object in the device storage_manager.
+///
+/// Arguments:
+/// mdoc_id: unique identifier for the credential to present, to be looked up
+///          in the VDC collection
+/// uuid:    the Bluetooth Low Energy Client Central Mode UUID to be used
+///
+/// Returns:
+/// A Result, with the `Ok` containing a tuple consisting of an enum representing
+/// the state of the presentation, a String containing the QR code URI, and a
+/// String containing the BLE ident.
+///
+use crate::common::*;
+use crate::{storage_manager::StorageManagerInterface, vdc_collection::VdcCollection};
+use std::ops::DerefMut;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use isomdl::{
+    definitions::{
+        device_engagement::{CentralClientMode, DeviceRetrievalMethods},
+        helpers::NonEmptyMap,
+        session, BleOptions, DeviceRetrievalMethod, SessionEstablishment,
+    },
+    presentation::device::{self, SessionManagerInit},
+};
+
+#[uniffi::export]
+pub async fn initialize_mdl_presentation(
+    uuid: Uuid,
+    storage_manager: Arc<dyn StorageManagerInterface>,
+) -> Result<MdlPresentationSession, SessionError> {
```

Comment:
As discussed out-of-band, we want to be able to initialize using the StorageManager, or using an mdoc directly, to enable multiple migration paths.